doc/BGP/BGP-router-id.md

```diff
@@ -0,0 +1,127 @@
+# BGP Router ID Explicitly Configured
+
+- [Revision](#revision)
+- [Definitions/Abbreviations](#definitionsabbreviations)
+- [Scope](#scope)
+- [Overview](#overview)
+- [Requirements](#requirements)
+- [High Level Design](#high-level-design)
+- [Config DB Enhancement](#config-db-enhancement)
+    - [DEVICE_METADATA](#device_metadata)
+
+### Revision
+
+| Revision | Date        | Author           | Change Description |
+| -------- | ----------- | ---------------- | ------------------ |
+| 1.0      | Mar 27 2024 | Yaqiang Zhu, Jing Kan | Initial proposal   |
+
+### Definitions/Abbreviations
+
+| Definitions/Abbreviation | Description |
+| ------------------------ | ----------- |
+| FRR | A free and open source Internet routing protocol suite for Linux and Unix platforms |
+| BGP Router ID | 32-bit value that uniquely identifies a BGP device |
+| AS | Autonomous System |
+
+### Scope
+
+This document describes a mechanism to allow user explicitly configure BGP router id.
+
+### Overview
+
+Currently, there are some BGP hard codings in SONiC:
+1. BGP router id was defined as a 32-bit value that uniquely identifies a BGP device. SONiC uses Loopback0 IPv4 address as BGP router id. This coupling prevents users from using customized router id. And FRR would choose an IP address in device to be BGP router id if Loopback0 IPv4 address doesn't exist. If the router id choosen by FRR is not unique in AS, it would be considered an error.
+2. SONiC wouldn't add BGP peer when there is not Loopback0 IPv4 exists. It would cause BGP cannot establish.
+
+Below is current workflow about BGP and router id, only includes contents related to Loopback0.
+
+1. After bgp container started, configuration file `/etc/frr/bgpd.conf` for bgpd would be rendered. It will use Loopback0 IPv4 address as BGP router id, if it doesn't exist, the BGP router id wouldn't be specified.
+2. bgpd start with configuration rendered before. If BGP router id is not specified, it would choose an IP address in device to be BGP router id.
+3. After bgpcfgd started, it will add bgp peer depends on whether Loopback0 IPv4 exist.
+
+<p align=center>
+<img src="img/origin_bgp_seq.png" alt="Figure 1. Origin bgp seq" width=700>
+</p>
+
+### Requirements
+
+Add support to allow user explicitly configure BGP router id.
+
+### High Level Design
+
+2 aspects enhancement:
+
+1. Add a field `bgp_router_id` in `CONFIG_DB["DEVICE_METADATA"]["localhost"]` to support explicitly configure BGP router id. If `CONFIG_DB["DEVICE_METADATA"]["localhost"]["bgp_router_id"]` configured, always use it as BGP router id no matter whether Loopback0 IPv4 exist. With this change, the new BGP router id configuration behavior will be like follow. To be clarified that when bgp_router_id doesn't be configured, the behavior is totally same as previously.
```

Comment:
In a multi-ASIC environment, would each ASIC have its own `bgp_router_id`?

Reply:
Yes